Make formspec elements real elements for draw order and clipping

DS-Minetest opened the pull request:
- Currently the formspec element draw order can't be changed for most elements. This is because they are drawn manually.
Also the clipping doesn't work correctly for many elements.
- This PR implements (nearly) all formspec elements as real gui elements and fixes those issues.
- The changes are taken from #8591 to make that PR simpler. #4844 is related but not fixed. Lists are still drawn at last because changing this is more difficult, but they are clippable.
Closes #8036.

## To do

This PR is a Ready for Review.

## How to test

<details><summary>Here is some lua code for testing.</summary>

```lua
-- edit this formspec,
-- especially change the order of the elements,
-- then compare. also compare to master

local my_formspec =
	"formspec_version[3]".. -- comment this to test backwards compatibility
	"size[10,8]"..
	"real_coordinates[true]"..

	"button[2.5,0.5;1,1;bla;Bla]"..
	"box[0,0;3,2;#00ff00ff]"..
	"tooltip[bla;Blatt;#f00;#000]"..
	"style[button_noclip;noclip=true]"..
	"button[-1,2.5;3,1;button_noclip;Noclip Button]"..
	"style[field_noclip;noclip=true]"..
	"field[-0.5,2.8;2,1;field_noclip;noclip field;some default text]"..
	"box[1,1;1,2;#1111ff]"..
	"bgcolor[;true]"..
	"background[1.5,1;2,2;default_wood.png;true]"..

	-- here are more elements, uncomment and test them
	--~ "scrollbar[7.5,0;0.3,4;vertical;scrbar;0]"..
	--~ "pwdfield[2,2;1,1;baum2;Baum]"..
	--~ "list[current_player;main;4,4;1,5;]"..
	--~ "image[3,1;bubble.png]"..
	--~ "item_image[2,6;3,2;default:mese]"..
	--~ "label[2,15;bla baum\nfoo bar]"..
	--~ "textarea[2,10;5,6;txtarea;Baum;bla]"..
	--~ "item_image_button[2,3;1,1;default:dirt_with_grass;itemimagebutton;ItemImageButton]"..
	--~ "tooltip[0,11;3,2;Blatt;#f00;#000]".."box[0,11;3,2;#00ff00]"..
	--~ "checkbox[2,25;chb;ChB;false]",
	--~ "bgcolor[#aa0a]"..
	--~ "tablecolumns[text,bla]"..
	--~ "table[1,23;3,1;tbl;a,b,c,d;2]"..
	--~ "textlist[0,5;2,0.5;txtlst;a,b,c]"..
	--~ "dropdown[0,6;2;hmdrpdwn;apfel,birne;1]"..
	--~ "tabheader[-1,9;mytabheader;a,b,c;2;false;false]"..
	--~ "background[5,5;1,1;testbg.png;true;10]"..
	""


minetest.register_on_player_receive_fields(function(player, formname, fields)
	if formname ~= "my_formspec" then
		print("other formname: "..formname)
		return
	end
	print(dump(fields))
end)


minetest.register_node("test:node", {
	description = "test node",
	tiles = {"default_wood.png^heart.png"},
	groups = {choppy = 3, oddly_breakable_by_hand = 3},
	on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
		minetest.show_formspec(clicker:get_player_name(), "my_formspec", my_formspec)
	end,
})
```
</details>

(Note that you might also want to test using `style`.)

rubenwardy commented:
The how to test section is intended to be used to show test code. Please test using a mod which demonstrates reordering

DS-Minetest commented:
Sorry, done.

DS-Minetest commented:
Rebased.

red-001 commented on src/gui/guiItemImage.h:
```diff
@@ -0,0 +1,47 @@
+
+	virtual void setText(const wchar_t *text)
+	{
+		m_label = core::stringw();
```
`irr::core::string` seems a have a assignment operator for c-style strings so you should be able to just do
```suggestion
		m_label = text;
```

DS-Minetest replied:
Thanks!
(I wonder, why I did that. I thought, this was copied from somewhere, but no.)

DS-Minetest commented on src/client/guiscalingfilter.h:
```diff
@@ -53,4 +53,5 @@ void draw2DImageFilterScaled(video::IVideoDriver *driver, video::ITexture *txr,
 void draw2DImage9Slice(video::IVideoDriver *driver, video::ITexture *texture,
-		const core::rect<s32> &rect, const core::rect<s32> &middle);
+		const core::rect<s32> &rect, const core::rect<s32> &middle,
+		const core::rect<s32> *cliprect = 0);
```
Done.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1806,35 +1913,45 @@ void GUIFormSpecMenu::parseItemImageButton(parserData* data, const std::string &
+			258 + m_fields.size()
+		);
+
+		GUIItemImage *e2 = new GUIItemImage(Environment, e, speci.fid,
```
bad name, use `e_img` or smth

DS-Minetest replied:
Done.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1806,35 +1913,45 @@ void GUIFormSpecMenu::parseItemImageButton(parserData* data, const std::string &
 		);
 
-		gui::IGUIButton *e = Environment->addButton(rect, this, spec.fid, L"");
+		gui::IGUIButton *e = Environment->addButton(rect, this, specb.fid, L"");
```
`e_btn`

DS-Minetest replied:
Done.

rubenwardy commented:
Buttons have disappeared from the scrollbar:

![image](https://user-images.githubusercontent.com/2122943/62706364-45bb5680-b9e7-11e9-867d-b1a720fde47f.png)


(this is with /formspec on minimal dev test)

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -507,7 +534,8 @@ void GUIFormSpecMenu::parseCheckbox(parserData* data, const std::string &element
 		}
 
-		m_checkboxes.emplace_back(spec,e);
+		e->grab();
```
Because of the complaints of stu, all pointers in `m_checkboxes`, `m_inventorylists` and co. are dropped when these containers are cleared. The `IGUICheckBox *` was not made with `new` but with an `add…` function. All pointers in `m_checkboxes` and co. are not only grabbed by the environment but also by the `GUIFormSpecMenu` code.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -641,17 +707,29 @@ void GUIFormSpecMenu::parseItemImage(parserData* data, const std::string &elemen
+
+		GUIItemImage *e = new GUIItemImage(Environment, this, spec.fid,
+				core::rect<s32>(pos, pos + geom), name,	m_font, m_client);
+		e->drop();
```
Yes, that's the reason.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1046,14 +1145,14 @@ void GUIFormSpecMenu::parseDropDown(parserData* data, const std::string &element
 
 		//now really show list
-		gui::IGUIComboBox *e = Environment->addComboBox(rect, this,spec.fid);
+		gui::IGUIComboBox *e = Environment->addComboBox(rect, this, spec.fid);
```
It is not made with `new` or `create…`, but with `add…`. See http://irrlicht.sourceforge.net/docu/classirr_1_1_i_reference_counted.html#details.

DS-Minetest commented:
(Huh, I didn't know about that chatcommand. >_<)

The error page seems to be Krock's fault, see https://github.com/minetest/minetest/pull/8731#issuecomment-519588432. (I wonder how he did that.)

DS-Minetest commented on src/gui/guiBackgroundImage.cpp:
```diff
@@ -0,0 +1,71 @@
+/*
+Minetest
+Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
```
Actually, a big part of the code is from you, too. (The 9-slice background.)

I've copied from `guiVolumeChange.h`, done.

DS-Minetest commented on src/gui/guiBackgroundImage.h:
```diff
@@ -0,0 +1,39 @@
+/*
+Minetest
+Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
```
Done

DS-Minetest commented:
The scrollbar buttons aren't there because they are set to not visible here:
https://github.com/minetest/minetest/blob/d1c27c7e8037e9f438741ee5f7d3b6bda22d22a0/src/gui/guiScrollBar.cpp#L422-L424
with this `border_size`:
https://github.com/minetest/minetest/blob/d1c27c7e8037e9f438741ee5f7d3b6bda22d22a0/src/gui/guiScrollBar.cpp#L337
(The scrollbar is not long enough relative to it's wideness.)

This will happen in other PRs that use the new scrollbar, too, like #8530.

Related: #8507 
>Notice in the second screenshot that the up/down arrow buttons are now automatically hidden when there is not enough space to fully accommodate them.

DS-Minetest commented:
`box` is the one element that commonly overlaps, I could use [`sendToBack`](http://irrlicht.sourceforge.net/docu/classirr_1_1gui_1_1_i_g_u_i_element.html#a9b1880048e7542dbe289295de69e77cd) on boxes if there's not `real_coordinates`.

DS-Minetest commented:
Rebased, the error window error is fixed.

rubenwardy commented:
https://forum.minetest.net/viewtopic.php?t=23286

Linuxdirk commented:
Just a quick question. Does it fix this?

https://github.com/minetest/minetest/issues/8036

rubenwardy commented:
> Just a quick question. Does it fix this?

Yes, for all elements except inventory list items, as the actual rendering of items in inventory lists is done last

DS-Minetest commented:
As #8036 is probably mostly fixed after this PR (except of with combination with inventory lists, which will make it a duplicate of #4844), I'll add an automatic close thing to this PR.

Linuxdirk commented:
> mostly fixed after this PR (except of with combination with inventory lists

Since the combination with inventory lists is the only thing mentioned in said issue it is not fixed at all then? :smile:

DS-Minetest commented:
> > mostly fixed after this PR (except of with combination with inventory lists
> 
> Since the combination with inventory lists is the only thing mentioned in said issue it is not fixed at all then? smile

Well, the title says "behind other elements".
That inventory lists are drawn first is already covered in #8036. Please, don't say, that your issue is a duplicate, we need to reach the status of being a game with a thousand issues. :slightly_smiling_face:

DS-Minetest commented:
Rebased and made this a non-breaking change.
I've added a new function `GUIFormSpecMenu::legacySortElements` to sort the elements into old order (for fs version < 2).

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2648,16 +2802,27 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 		// Backup the coordinates so that prepends can use the coordinates of choice.
 		bool rc_backup = mydata.real_coordinates;
-		bool version_backup = m_formspec_version;
+		u16 version_backup = m_formspec_version;
```
this should be fixed on master before this PR is merged

DS-Minetest replied:
Done.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2707,6 +2872,66 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
+	 * 2. all normal elements (like buttons)
+	 * 3. images
+	 * 4. item images
+	 * 5. item lists (TODO)
```
Is this needed given that the new system draws inv lists in front anyway? This would explain why this is todo

DS-Minetest replied:
Yes, the lists are still drawn in front with this PR. But this should be fixed at some point, hence the todo.

rubenwardy commented:
Segfault in main menu

```
0x00007ffff66ddda8 in __cxxabiv1::__dynamic_cast (src_ptr=0x5555564cede0, 
    src_type=0x555555d68438 <typeinfo for irr::gui::IGUIElement>, 
    dst_type=0x555555d6a300 <typeinfo for GUIBox>, src2dst=0)
    at /build/gcc/src/gcc/libstdc++-v3/libsupc++/dyncast.cc:68
68	/build/gcc/src/gcc/libstdc++-v3/libsupc++/dyncast.cc: No such file or directory.
(gdb) bt full
#0  0x00007ffff66ddda8 in __cxxabiv1::__dynamic_cast (src_ptr=0x5555564cede0, 
    src_type=0x555555d68438 <typeinfo for irr::gui::IGUIElement>, 
    dst_type=0x555555d6a300 <typeinfo for GUIBox>, src2dst=0)
    at /build/gcc/src/gcc/libstdc++-v3/libsupc++/dyncast.cc:68
        vtable = 0x7ffff7d566a8 <vtable for irr::gui::CGUITabControl+32>
        prefix = 0x7ffff7d56698 <vtable for irr::gui::CGUITabControl+16>
        whole_ptr = 0x5555564cede0
        whole_type = 0x0
        result = {dst_ptr = 0x0, 
          whole2dst = __cxxabiv1::__class_type_info::__unknown, 
          whole2src = __cxxabiv1::__class_type_info::__unknown, 
          dst2src = __cxxabiv1::__class_type_info::__unknown, 
          whole_details = 16}
        whole_vtable = 0x7ffff7d566a8 <vtable for irr::gui::CGUITabControl+32>
        whole_prefix = 0x7ffff7d56698 <vtable for irr::gui::CGUITabControl+16>
#1  0x000055555587df10 in GUIFormSpecMenu::legacySortElements(irr::core::list<irr::gui::IGUIElement*>::Iterator) ()
No symbol table info available.
#2  0x00005555558a2a57 in GUIFormSpecMenu::regenerateGui(irr::core::vector2d<unsigned int>) ()
No symbol table info available.
#3  0x0000555555896f86 in GUIFormSpecMenu::drawMenu() ()
No symbol table info available.
--Type <RET> for more, q to quit, c to continue without paging--
#4  0x00005555558ef128 in GUIModalMenu::draw() ()
No symbol table info available.
#5  0x00007ffff7bc7142 in irr::gui::CGUIStaticText::draw() () from /usr/lib/libIrrlicht.so.1.8
No symbol table info available.
#6  0x00007ffff7b73d5a in irr::gui::CGUIEnvironment::drawAll() () from /usr/lib/libIrrlicht.so.1.8
No symbol table info available.
#7  0x000055555587afb1 in GUIEngine::run() ()
No symbol table info available.
#8  0x000055555587bb30 in GUIEngine::GUIEngine(JoystickController*, irr::gui::IGUIElement*, IMenuManager*, MainMenuData*, bool&) ()
No symbol table info available.
#9  0x00005555557208b8 in ClientLauncher::main_menu(MainMenuData*) ()
No symbol table info available.
#10 0x00005555557217b3 in ClientLauncher::launch_game(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, bool, GameParams&, Settings const&) ()
No symbol table info available.
#11 0x0000555555723d12 in ClientLauncher::run(GameParams&, Settings const&) ()
No symbol table info available.
#12 0x00005555556f85e9 in main ()
No symbol table info available.
```

DS-Minetest commented:
The rect of inventory lists is now calculated correctly:
![screenshot_20190915_185104](https://user-images.githubusercontent.com/7613443/64924965-5cc14580-d7ea-11e9-9f43-acf2ce1b7f9a.png)
Thanks for testing!

rubenwardy commented:
```
Thread 1 "minetest" received signal SIGSEGV, Segmentation fault.
irr::core::list<irr::gui::IGUIElement*>::Iterator::operator++ (this=0x7fffffffc9f8) at /usr/include/irrlicht/irrList.h:45
45			Iterator  operator ++(s32) { Iterator tmp = *this; Current = Current->Next; return tmp; }
(gdb) bt full
#0  irr::core::list<irr::gui::IGUIElement*>::Iterator::operator++ (this=0x7fffffffc9f8) at /usr/include/irrlicht/irrList.h:45
        tmp = {Current = 0x0}
#1  0x0000555555d60968 in GUIFormSpecMenu::regenerateGui (this=0x555556b402d0, screensize=...) at /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:2878
        mydata = {explicit_size = false, real_coordinates = false, invsize = {X = 0, Y = 0}, size = {X = 100, Y = 100}, offset = {X = 0.5, Y = 0.5}, anchor = {X = 0.5, Y = 0.5}, rect = {
            UpperLeftCorner = {X = 0, Y = 0}, LowerRightCorner = {X = 0, Y = 0}}, basepos = {X = 139, Y = 145}, screensize = {X = 858, Y = 591}, focused_fieldname = "", 
          table_options = std::vector of length 0, capacity 0, table_columns = std::vector of length 0, capacity 0, table_dyndata = std::unordered_map with 0 elements}
        focused_element = 0x555556b402d0
        __PRETTY_FUNCTION__ = "virtual void GUIFormSpecMenu::regenerateGui(v2u32)"
        elements = std::vector of length 1, capacity 1 = {""}
        i = 1
        enable_prepends = true
        skin = 0x55555694c180
        old_font = 0x555556b295a0
        legacy_sort_start = {Current = 0x0}
#2  0x0000555555da9749 in GUIModalMenu::draw (this=0x555556b402d0) at /home/ruben/dev/minetest/src/gui/modalMenu.cpp:76
        driver = 0x5555568d6520
        screensize = {X = 858, Y = 591}
#3  0x00007ffff7bc7142 in irr::gui::CGUIStaticText::draw() () from /usr/lib/libIrrlicht.so.1.8
No symbol table info available.
#4  0x00007ffff7b73d5a in irr::gui::CGUIEnvironment::drawAll() () from /usr/lib/libIrrlicht.so.1.8
No symbol table info available.
#5  0x0000555555d47390 in GUIEngine::run (this=0x7fffffffd070) at /home/ruben/dev/minetest/src/gui/guiEngine.cpp:296
        current_screen_size = @0x5555568d6620: {Width = 858, Height = 591}
        driver = 0x5555568d6520
        text_height = 20
        previous_screen_size = {Width = 858, Height = 591}
        sky_color = {color = 4287412986}
#6  0x0000555555d467be in GUIEngine::GUIEngine (this=0x7fffffffd070, joystick=0x555556b268c8, parent=0x555556a30f10, menumgr=0x55555646fc60 <g_menumgr>, data=0x7fffffffd310, 
    kill=@0x5555564a2800: false) at /home/ruben/dev/minetest/src/gui/guiEngine.cpp:196
        soundfetcher = {<OnDemandSoundFetcher> = {_vptr.OnDemandSoundFetcher = 0x5555563ff668 <vtable for MenuMusicFetcher+16>}, m_fetched = std::set with 1 element = {[0] = "main_menu"}}
        rect = {UpperLeftCorner = {X = 4, Y = 0}, LowerRightCorner = {X = 4, Y = 20}}
        soundfetcher = <optimized out>
        rect = <optimized out>
        texture = <optimized out>
        __for_range = <optimized out>
        __for_begin = <optimized out>
        __for_end = <optimized out>
        e = <optimized out>
#7  0x0000555555bf2d6d in ClientLauncher::main_menu (this=0x7fffffffdbe0, menudata=0x7fffffffd310) at /home/ruben/dev/minetest/src/client/clientlauncher.cpp:560
        kill = 0x5555564a2800 <porting::g_killed>
        driver = 0x5555568d6520
        mymenu = {_vptr.GUIEngine = 0x5555563ff648 <vtable for GUIEngine+16>, m_parent = 0x555556a30f10, m_menumanager = 0x55555646fc60 <g_menumgr>, m_smgr = 0x555556b1b0f0, 
          m_data = 0x7fffffffd310, m_texture_source = 0x555556b3fc50, m_sound_manager = 0x555556b400e0, m_formspecgui = 0x555556b3fbd0, m_buttonhandler = 0x555556b400a0, 
          m_menu = 0x555556b402d0, m_kill = @0x5555564a2800, m_startgame = false, m_script = 0x555556b3a430, m_scriptdir = "/home/ruben/dev/minetest/bin/../builtin/mainmenu", m_textures = {
            {texture = 0x0, tile = false, minsize = 0}, {texture = 0x0, tile = 128, minsize = 32767}, {texture = 0x555556b28b90, tile = false, minsize = 16}, {texture = 0x0, tile = 192, 
              minsize = 32767}}, m_irr_toplefttext = 0x555556a092b0, m_toplefttext = {m_string = L"Minetest Game", m_colors = std::vector of length 13, capacity 16 = {{color = 4294967295}, 
              {color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {
                color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {color = 4294967295}, {color = 4294967295}}, m_has_background = false, m_background = {color = 0}}, 
          m_clouds_enabled = true, m_cloud = {dtime = 0.0340000018, lasttime = 193, clouds = 0x555556509a90, camera = 0x5555569d80d0}}
#8  0x0000555555bf1e00 in ClientLauncher::launch_game (this=0x7fffffffdbe0, error_message="", reconnect_requested=false, game_params=..., cmd_args=...)
    at /home/ruben/dev/minetest/src/client/clientlauncher.cpp:424
        newport = 32767
        worldspecs = std::vector of length -35394917066724837, capacity -87413170238280657 = {<error reading variable worldspecs (Cannot access memory at address 0x74752e42475f6e65)>
        menudata = {servername = "", serverdescription = "", address = "", port = "30000", name = "rubenwardy", password = "", do_reconnect = false, selected_world = 0, 
          simple_singleplayer_mode = false, script_data = {reconnect_requested = false, errormessage = ""}}
--Type <RET> for more, q to quit, c to continue without paging--
        __PRETTY_FUNCTION__ = "bool ClientLauncher::launch_game(std::string&, bool, GameParams&, const Settings&)"
#9  0x0000555555bf0654 in ClientLauncher::run (this=0x7fffffffdbe0, game_params=..., cmd_args=...) at /home/ruben/dev/minetest/src/client/clientlauncher.cpp:218
        game_has_run = false
        text = 0x555556b3ffb0 L"\x55ee4310啕\x55eb2506啕\x55ee05e8啕\x61754c\x656c61\060"
        __PRETTY_FUNCTION__ = "bool ClientLauncher::run(GameParams&, const Settings&)"
        camera = 0x555556b33d40
        chat_backend = {m_console_buffer = {m_scrollback = 500, m_unformatted = std::vector of length 0, capacity 0, m_cols = 0, m_rows = 0, m_scroll = 0, 
            m_formatted = std::vector of length 0, capacity 0, m_empty_formatted_line = {fragments = std::vector of length 0, capacity 0, first = true}}, m_recent_buffer = {
            m_scrollback = 6, m_unformatted = std::vector of length 0, capacity 0, m_cols = 0, m_rows = 0, m_scroll = 0, m_formatted = std::vector of length 0, capacity 0, 
            m_empty_formatted_line = {fragments = std::vector of length 0, capacity 0, first = true}}, m_prompt = {m_prompt = L"]", m_line = L"", 
            m_history = std::vector of length 0, capacity 0, m_history_index = 0, m_history_limit = 500, m_cols = 0, m_view = 0, m_cursor = 0, m_cursor_len = 0, 
            m_nick_completion_start = 0, m_nick_completion_end = 0}}
        error_message = ""
        reconnect_requested = false
        first_loop = true
        retval = true
        kill = 0x5555564a2800 <porting::g_killed>
#10 0x000055555602230b in main (argc=1, argv=0x7fffffffdf58) at /home/ruben/dev/minetest/src/main.cpp:217
        retval = 21845
        cmd_args = {m_settings = std::unordered_map with 0 elements, m_defaults = std::unordered_map with 0 elements, m_callbacks = std::unordered_map with 0 elements, 
          m_callback_mutex = {<std::__mutex_base> = {_M_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {
                    __prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}}, <No data fields>}, m_mutex = {<std::__mutex_base> = {_M_mutex = {__data = {__lock = 0, 
                  __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, 
                __align = 0}}, <No data fields>}}
        cmd_args_ok = true
        game_params = {socket_port = 30000, world_path = "/home/ruben/dev/minetest/bin/../worlds/world", game_spec = {id = "minetest", name = "Minetest Game", author = "Minetest", 
            release = 0, path = "/home/ruben/dev/minetest/bin/../games/minetest_game", gamemods_path = "/home/ruben/dev/minetest/bin/../games/minetest_game/mods", 
            addon_mods_paths = std::set with 1 element = {[0] = "/home/ruben/dev/minetest/bin/../mods"}, 
            menuicon_path = "/home/ruben/dev/minetest/bin/../games/minetest_game/menu/icon.png"}, is_dedicated_server = false}
        isServer = false
        __PRETTY_FUNCTION__ = "int main(int, char**)"
        launcher = {list_video_modes = false, skip_main_menu = false, use_freetype = true, random_input = false, address = "", playername = "rubenwardy", password = "", 
          input = 0x555556b268c0, receiver = 0x7fffe800bbd0, skin = 0x55555694c180, font = 0x0, gamespec = {id = "", name = "", author = "", release = 0, path = "", gamemods_path = "", 
            addon_mods_paths = std::set with 0 elements, menuicon_path = ""}, worldspec = {path = "/home/ruben/dev/minetest/bin/../worlds/world", name = "[--world parameter] [new]", 
            gameid = "minetest"}, simple_singleplayer_mode = false, current_playername = "inv£lid", current_password = "", current_address = "does-not-exist", current_port = 0}
```

DS-Minetest commented:
The bugs should be fixed now.

jastevenson303 commented:
Tested with Lavaland, quickly.

Noticed that I was unable to use the [List] [Ok] buttons or type a name in the field on this formspec (beds)
![screenshot_20190916_045511](https://user-images.githubusercontent.com/23448162/64945338-52c93200-d83e-11e9-8e94-29993c57b553.png)  The other buttons and checkbox work, however.

Formspecs look different?  A regular field:
![screenshot_20190916_045451](https://user-images.githubusercontent.com/23448162/64945353-5a88d680-d83e-11e9-85af-3173e05c6fa0.png)

ClobberXD commented on src/gui/guiBackgroundImage.h:
```diff
@@ -0,0 +1,37 @@
+/*
+Part of Minetest
```
Why `part of`? Other files just have `Minetest`.

DS-Minetest replied:
Afaik, it means that the file is part of minetest.

See also https://github.com/minetest/minetest/pull/8740#discussion_r312014639.

ClobberXD replied:
All files are part of Minetest tho. Also, the linked comment says nothing about adding "Part of"

ClobberXD commented on src/gui/guiBox.cpp:
```diff
@@ -0,0 +1,38 @@
+		return;
+
+	Environment->getVideoDriver()->draw2DRectangle(
+			m_color, AbsoluteRect, &AbsoluteClippingRect);
```
Extra indent

DS-Minetest replied:
Afaik, it's allowed.

ClobberXD commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -395,7 +394,35 @@ void GUIFormSpecMenu::parseList(parserData* data, const std::string &element)
+			"",
+			L"",
+			L"",
+			258 + m_fields.size()
```
Couldn't these be placed in one line?

DS-Minetest replied:
They could, but all field specs are created in this format. Hence changing one would require changing all, making this PR even larger.

ClobberXD replied:
Understood; I noticed that too.

DS-Minetest commented:
Thanks for testing, @jastevenson303!
The first bug should actually already be fixed by https://github.com/minetest/minetest/pull/8740/commits/6055a5f7b32c36232e49d366c09fb5bb169d9fc9. Did you test with the newest version of this PR?
I can reproduce the second thing and will look into it.

ClobberXD commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1235,21 +1338,26 @@ void GUIFormSpecMenu::createTextField(parserData *data, FieldSpec &spec,
+
+		if (t) {
+			t->setNotClipped(style.getBool(StyleSpec::NOCLIP, false));
+		}
```
Braces :D

DS-Minetest replied:
Done.

ClobberXD commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2760,9 +2991,9 @@ GUIFormSpecMenu::ItemSpec GUIFormSpecMenu::getItemAtPos(v2s32 p) const
-			{
+			core::rect<s32> rect = imgrect + base_pos + p0;
+			rect.clipAgainst(clipping_rect);
+			if (rect.getArea() > 0 && rect.isPointInside(p)) {
```
Braces :D

DS-Minetest replied:
Done.

DS-Minetest commented:
The simple field is now fixed.

jastevenson303 commented:
This is with ad4b58dd93ae2bd9dcf7786bcb3e2c57bdf48659 @DS-Minetest

DS-Minetest commented:
@jastevenson303 I think, the problem is that you have set the height to 4 but the inventory only has a size to allow a height of 3. Therefor the list overlaps with invisible slots.
I'd suggest to change the 4 [here](https://github.com/jastevenson303/Lavaland/blob/f06266f693e07f8f8a732e62442af06c2230d1eb/mods/beds/api.lua#L142) to a 3.

I'm not sure how I could fix such things.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3346,7 +3468,7 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
 					// no dynamic cast possible due to some distributions shipped
 					// without rttzi support in irrlicht
-					IGUIElement * element = getElementFromId(s.fid);
+					IGUIElement * element = getElementFromId(s.fid, true);
```
Done.
(Github can be weird sometimes. Why does it show the number 3471 up there? My editor says 3470 (and it starts counting at 1). The change suggestion also says 3470:
![2019-09-16-183836_1920x1080_scrot](https://user-images.githubusercontent.com/7613443/64976400-76799000-d8b1-11e9-94fa-072cedb47cd4.png))

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3361,7 +3483,7 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
 					// no dynamic cast possible due to some distributions shipped
 					// without rtti support in irrlicht
-					IGUIElement * element = getElementFromId(s.fid);
+					IGUIElement * element = getElementFromId(s.fid, true);
```
Done.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3377,13 +3499,12 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
-					}
+					IGUIElement *element = getElementFromId(s.fid, true);
+					GUIScrollBar *e = nullptr;
+					if ((element) && (element->getType() == gui::EGUIET_ELEMENT))
```
Done.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -4176,6 +4295,17 @@ std::string GUIFormSpecMenu::getNameByID(s32 id)
+FormspecFieldType GUIFormSpecMenu::getTypeByID(s32 id)
+{
+	for (FieldSpec &spec : m_fields) {
+		if (spec.fid == id) {
```
The function was copied from `getNameByID`. I've removed braces there and in the other function, too.
I've kept the outer `for` braces:
>Also, use braces when a loop contains more than a single simple statement:

https://www.kernel.org/doc/html/latest/process/coding-style.html#placing-braces-and-spaces

ClobberXD replied:
Even if there's only one single statement, we generally always use braces for loops, I think. So that's not a problem :)

DS-Minetest commented on src/network/networkprotocol.h:
```diff
@@ -228,6 +228,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 	FORMSPEC VERSION 2:
 		Forced real coordinates
 		background[]: 9-slice scaling parameters
+		Formspec elements are drawn in correct order
```
Done.

DS-Minetest commented on src/client/guiscalingfilter.cpp:
```diff
@@ -224,7 +225,7 @@ void draw2DImage9Slice(video::IVideoDriver *driver, video::ITexture *texture,
 			draw2DImageFilterScaled(driver, texture, dest,
 					src,
-					NULL/*&AbsoluteClippingRect*/, colors, true);
+					cliprect, colors, true);
```
When I see something weird, I always think that there's a reason behind it, else it wouldn't be done like this. Therefor I had not changed this.
Done.

DS-Minetest commented on src/gui/guiBackgroundImage.cpp:
```diff
@@ -0,0 +1,69 @@
+				core::rect<s32>(core::position2d<s32>(0, 0),
+						core::dimension2di(
+								texture->getOriginalSize())),
+				&AbsoluteClippingRect, colors, true);
```
I've tried my best. Done.

ClobberXD replied:
No pressures, thanks.

DS-Minetest commented on src/gui/guiBackgroundImage.cpp:
```diff
@@ -0,0 +1,69 @@
+		// `-x` is interpreted as `w - x`
+		if (middle.LowerRightCorner.X < 0) {
+			middle.LowerRightCorner.X += texture->getOriginalSize().Width;
+		}
```
Done.

DS-Minetest commented on src/gui/guiBackgroundImage.cpp:
```diff
@@ -0,0 +1,69 @@
+			middle.LowerRightCorner.Y += texture->getOriginalSize().Height;
+		}
+		draw2DImage9Slice(driver, texture, AbsoluteRect, middle,
+				&AbsoluteClippingRect);
```
Afaik, there's no rule that forbids this.
And it looks better (= more readable) with two indents because of the two `if` statements above (less monotonous).

ClobberXD replied:
If you think that's more readable, no problemo, thanks

DS-Minetest commented:
Thanks for the code style review @ClobberXD!

jastevenson303 commented:
Haha, that's funny, so you're saying your PR exposed a bug in my formspec
that the old version didn't show :D  Congrats!

On Mon, Sep 16, 2019 at 1:11 PM DS <notifications@github.com> wrote:

> Thanks for the code style review @ClobberXD <https://github.com/ClobberXD>
> !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/8740?email_source=notifications&email_token=AFS4UYW3OS67OXLV3BEEC7DQJ6427A5CNFSM4II7EJIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6Z3CJY#issuecomment-531869991>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFS4UYS2G5CVYSUXM5B6LYLQJ6427ANCNFSM4II7EJIA>
> .
>

DS-Minetest commented:
Ok, I've fixed your bug, @jastevenson303.
In your particular situation, it is a bug caused by you, evil you. But there might be many people who do the same mistake (because of lazy copying). Because of this I've fixed this.
Also, now there won't be warnings anymore in the draw function if the inventory or the list does not exist (in most cases; there might still be warnings if the inventory somehow disappears). (The warnings were like torturous screams of horror.)

DS-Minetest commented:
Thanks!

(The last commit has no functional changes.)

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2707,6 +2900,68 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
+	for (auto i = from; i != to;) {
+		if (getTypeByID((*i)->getID()) == f_Box) {
+			auto e = *i;
+			i = Children.erase(i);
```
`std::sort` could do this. You'd only need a function to get the priority, and one to compare one element's priority with another.

Example:
https://github.com/minetest/minetest/pull/8603/files#diff-6ba6d577c564534741b34ee55c76e5b8R343
https://github.com/minetest/minetest/pull/8603/files#diff-6ba6d577c564534741b34ee55c76e5b8R446

Just for your information. I think this code will work good as well.

DS-Minetest replied:
Would that ensure that the order otherwise stays the same? Eg. `button[1]box[1]image[1]button[2]box[2]button[3]image[2]` has to be `box[1]box[2]button[1]button[2]button[3]image[1]image[2]` and the internal order of boxes, buttons or images here has to be 123.

SmallJoker replied:
Same elements have the same priority, hence their relative positions never change.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3392,12 +3538,10 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
-				{
-					IGUIElement* e = getElementFromId(s.fid);
-					if(e != NULL) {
+				else {
```
`} else {`

DS-Minetest replied:
This is adapted to the code style of the whole if else construct here. I would have to change it everywhere.
(Also it's quite readable.)

SmallJoker replied:
Only needed for touched lines

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3392,12 +3538,10 @@ void GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode=quit_mode_no)
-					if(e != NULL) {
+				else {
+					IGUIElement *e = getElementFromId(s.fid, true);
+					if(e)
```
Space

DS-Minetest replied:
Done.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -4169,13 +4313,22 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 }
 
+
+FormspecFieldType GUIFormSpecMenu::getTypeByID(s32 id)
```
This function could be merged with `getLabelByID`: `getFieldSpecById`, returning the entire fieldspec (or a default one).

DS-Minetest replied:
`getLabelByID` is inherited from `GUIModalMenu`:
https://github.com/minetest/minetest/blob/9519d57017cfa28eb4f2fbf3c780f826017bbb00/src/gui/modalMenu.h#L62-L63
See also: https://github.com/minetest/minetest/search?q=getLabelByID&unscoped_q=getLabelByID

SmallJoker replied:
:man_shrugging: well then, OK.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -812,8 +919,31 @@ void GUIFormSpecMenu::parseBackground(parserData* data, const std::string &eleme
+		);
+
+		core::rect<s32> rect;
+		if (!clip)
```
Done.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1807,35 +1939,45 @@ void GUIFormSpecMenu::parseItemImageButton(parserData* data, const std::string &
 
-		FieldSpec spec(
+		// the spec for the button
+		FieldSpec specb(
```
Done.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1807,35 +1939,45 @@ void GUIFormSpecMenu::parseItemImageButton(parserData* data, const std::string &
-		else
-			pos = getElementBasePos(true, &v_pos);
+		// the spec for the item-image
+		FieldSpec speci(
```
Done.

DS-Minetest commented:
Rebased.

DS-Minetest commented:
Luckily the `bgcolor` fix PR's drawback has shown me a bug here. Fixed.
(The size of the rect of the form is not given in prepends.)

DS-Minetest commented:
`override` added. See also #8957.

DS-Minetest commented:
Why was this removed from the milestone?

Most of the complaints of travis and its suggestions don't meet the code style guidelines.

rubenwardy commented:
It may still be merged if you pester Krock. But it's not a blocker and the feature freeze is in 5 days

SmallJoker commented:
Seen, noted. I don't see an urgent need for this PR to be in 5.1.0. The code is good and works (will do some testing, though), but waiting until after release will give some more time to find & fix edge-cases, if there are any.

rubenwardy commented:
That is also true. This is a relatively risky PR given its nature

DS-Minetest commented:
Rebased.

SmallJoker commented:
Quick update: Tested the PR with various mods, and also on a remote server with different formspecs. Could yet not find any visual difference, but until I feel safe to give an approval I'll study the code in the next few days. Interesting that you could get rid of the absolute position calculations.

DS-Minetest commented:
I guess I should bump the formspec version, right?

SmallJoker commented:
@DS-Minetest Yes, that would be great if you could make this depend on version 3.

DS-Minetest commented:
Done.

SmallJoker commented:
Minetest freezes when there's an error from Lua:
```
Thread 1 "minetest" received signal SIGINT, Interrupt.
0x000055555582a04e in GUIFormSpecMenu::legacySortElements(irr::core::list<irr::gui::IGUIElement*>::Iterator) ()
(gdb) bt
#0  0x000055555582a04e in GUIFormSpecMenu::legacySortElements(irr::core::list<irr::gui::IGUIElement*>::Iterator) ()
#1  0x000055555584d926 in GUIFormSpecMenu::regenerateGui(irr::core::vector2d<unsigned int>) ()
#2  0x000055555584267d in GUIFormSpecMenu::drawMenu() ()
#3  0x00005555558a3024 in GUIModalMenu::draw() ()
#4  0x00007ffff78afce2 in irr::gui::CGUIStaticText::draw() ()
   from /usr/lib/x86_64-linux-gnu/libIrrlicht.so.1.8
#5  0x00007ffff786472a in irr::gui::CGUIEnvironment::drawAll() ()
   from /usr/lib/x86_64-linux-gnu/libIrrlicht.so.1.8
```
Code used:
```Lua
) end
```

DS-Minetest commented:
Ugh, I can reproduce.
What does actually happen when there's a lua error?

DS-Minetest commented:
Fixed.

Inserting an element in front of a removed iterator caused the problem.

The crash actually just happened because the error window has a box in its back.

Thanks[,](http://irc.minetest.net/minetest-dev/2019-10-31#i_5607630) Krock!

DS-Minetest commented:
Rebased.

DS-Minetest commented:
Yey! :tada:
